R/imageutils.R

```diff
@@ -50,9 +50,12 @@ plotPNG <- function(func, filename=tempfile(fileext='.png'),
   # guarantee user's code in func() will not trigger the error -- they may have
   # to set par(mar = smaller_value) before they draw base graphics.
   op <- graphics::par(mar = rep(0, 4))
-  tryCatch(graphics::plot.new(), finally = graphics::par(op))
+  on.exit(graphics::par(op), add = TRUE)
+  graphics::plot.new()
+
   dv <- grDevices::dev.cur()
-  tryCatch(shinyCallingHandlers(func()), finally = grDevices::dev.off(dv))
+  on.exit(grDevices::dev.off(dv), add = TRUE)
+  func()
```

Comment:
The point of `plot.new()` is to guarantee at least a blank PNG when `renderPlot()` does not really contain code that generates a plot, but we need to make sure `plot.new()` always works. One situation in which it does not work is when the plot size is too small, and the error is "figure margin too large". The solution is set zero margins `par(mar = rep(0, 4))`. But we also have to make sure we do not let this `mar` setting affect user's plot, so we have to reset `par()` _immediately_ after `plot.new()`.

We can certainly use `on.exit()` if it looks cleaner, and we just need to introduce a helper function like

```r
blank_plot = function() {
  op = par(mar = rep(0, 4))
  on.exit(par(op), add = TRUE)
  plot.new()
}
```

Then we call `blank_plot()`.